[DFT][rocFFt] Address rocFFT failing tests

s-Nick opened the pull request:
# Description

This PR addresses two issues arose recently in rocFFT backend. Both issues are related to rocFFT internal changes.
One issue is fixed by small modification to oneMKL backend. The other issue cannot be fixed internally, but it was fixed by AMD and it will available in the future versions of rocFFT. 
To avoid that users incur with the second issues a warning is emitted during the configuration and eventually an exception is thrown. 

first issue -> https://github.com/ROCm/rocFFT/issues/504 
second issue -> https://github.com/ROCm/rocFFT/issues/507

Fixes # (GitHub issue)

This PR address oneMKL issue #559 

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
older rocm version(5.4.3) -> [rocm5_4_3_full_log.txt](https://github.com/user-attachments/files/16817642/rocm5_4_3_full_log.txt)
current release -> [rocm6_1_fix_log.txt](https://github.com/user-attachments/files/16817651/rocm6_1_fix_log.txt)
rocFFT develop branch -> [develop_fix_log.txt](https://github.com/user-attachments/files/16817657/develop_fix_log.txt)

- [x] Have you formatted the code using clang-format?

## Bug fixes

- [x] Have you added relevant regression tests?
- [x] Have you included information on how to reproduce the issue (either in a
      GitHub issue or in this PR)?
 issue #559

hjabird committed:
[rocFFT] Work around bug in rocFFT for ROCm 6

* rocFFT distributed with ROCm 6+ contains a bug meaning that the
strides cannot be set on a plan description twice.
* This commit works around this issue.

s-Nick committed:
[rocFFT] Add checks for rocFFT version due to internal bug

Due to rocFFt internal bug https://github.com/ROCm/rocFFT/issues/507
Add cmake version checks based upon rocFFT version.
Add exception to deal with faulty cases for affected rocFFT version.
RocFFT versions taken from https://github.com/ROCm/rocFFT/blob/develop/CHANGELOG.md

Rbiessy commented on src/dft/backends/rocfft/CMakeLists.txt:
```diff
@@ -49,6 +49,14 @@ find_package(HIP REQUIRED)
 # Require the minimum rocFFT version matching with ROCm 5.4.3.
 find_package(rocfft 1.0.21 REQUIRED)
 
+if (${rocfft_VERSION_MAJOR} EQUAL "1" AND ${rocfft_VERSION_MINOR} EQUAL "0"
+    AND ((${rocfft_VERSION_PATCH} GREATER "22")
+    AND (${rocfft_VERSION_PATCH} LESS "31")  ))
+    message(WARNING "Due to a bug in rocFFT some tests fail with the version in\
+      use. If possible use a version greater of 1.0.30 or less of 1.0.23.
+      Current rocFFT version ${rocfft_VERSION}")
+endif()
```
I don't think this is needed since users should see the unimplemented exception when using this rocFFT version, not the tests failing.

s-Nick replied:
Thank you for your review @Rbiessy . I added it to warn the user before compiling the library even without the test, but if you think it is redundant I can remove it.

Rbiessy replied:
Yeah it can be removed.

Rbiessy commented on src/dft/backends/rocfft/commit.cpp:
```diff
@@ -259,12 +260,37 @@ class rocfft_commit final : public dft::detail::commit_impl<prec, dom> {
-        rocfft_plan_description plan_desc;
-        if (rocfft_plan_description_create(&plan_desc) != rocfft_status_success) {
+        // This workaround is needed due to a confirmed issue in rocFFT from version
+        // 1.0.23 to 1.0.30. Those rocFFT version correspond to rocm version from
+        // 5.6.0 to 6.3.0.
```
Where did you get the information about ROCm 6.3.0 and rocFFT 1.0.30? The latest versions I can see are [ROCm 6.2.0](https://github.com/ROCm/ROCm/releases) which uses [rocFFT 1.0.28](https://rocm.docs.amd.com/en/docs-6.2.0/compatibility/compatibility-matrix.html). I think we can assume that the fix will be part of rocFFT 1.0.29 and ROCm 6.3.0.

s-Nick replied:
I took from their change log in the rocFFT repo https://github.com/ROCm/rocFFT/blob/develop/CHANGELOG.md
According to it, there should be a ROCM 6.2.1 that uses rocFFT 1.0.29 then ROCm 6.3.0 with rocFFT 1.0.30 and then the fix for one of the issues

Rbiessy replied:
Thanks that looks good then.

Rbiessy commented on src/dft/backends/rocfft/commit.cpp:
```diff
@@ -259,12 +260,37 @@ class rocfft_commit final : public dft::detail::commit_impl<prec, dom> {
+                    return true;
+                };
+                std::printf("hello\n");
+                if (!stride_checker(stride_vecs.vec_a, stride_vecs.vec_b))
```
Can't you use `stride_vecs.vec_a != stride_vecs.vec_b`? Or is `dimensions` smaller than the length of these vectors?

s-Nick replied:
Honestly, I did not check. Dimensions should be the same. I'll try and update it.

s-Nick committed:
Remove debug printing point

s-Nick commented on src/dft/backends/rocfft/commit.cpp:
```diff
@@ -39,6 +39,7 @@
 #include "rocfft_handle.hpp"
 
 #include <rocfft.h>
+#include <rocfft-version.h>
```
This header is now needed to check rocFFT version if necessary

Rbiessy commented on src/dft/backends/rocfft/commit.cpp:
```diff
@@ -259,12 +260,37 @@ class rocfft_commit final : public dft::detail::commit_impl<prec, dom> {
+                if (!stride_checker(stride_vecs.vec_a, stride_vecs.vec_b))
+                    throw oneapi::mkl::unimplemented(
+                        "DFT", func,
+                        "due to a bug in rocfft version in use, it requires fwd and bwd stride to be the same for COMPLEX out_of_place computations");
```
I think this whole check should be run in the out-of-place compute functions otherwise it looks like this will prevent users from running complex in-place FFTs for no reasons.

s-Nick replied:
yes, that would make more sense, but we can let that function clean and I can add another check in the `if` at [line 269](https://github.com/oneapi-src/oneMKL/blob/54e0cf91910a6584e8c81c522f0a2417cb7a6e10/src/dft/backends/rocfft/commit.cpp#L269). I added a comment directly to that line.

If you think change the compute function it is a better choice I am ok with it too.

Rbiessy replied:
You're right, I forgot that the placement was already known at commit time so that makes. Btw you should make sure the changes are formatted with clang-format-9 to avoid issues in the CI!

s-Nick commented on src/dft/backends/rocfft/commit.cpp:
```diff
@@ -259,12 +260,36 @@ class rocfft_commit final : public dft::detail::commit_impl<prec, dom> {
+        // 1.0.23 to 1.0.30. Those rocFFT version correspond to rocm version from
+        // 5.6.0 to 6.3.0.
+        // Link to rocFFT issue: https://github.com/ROCm/rocFFT/issues/507
+        if constexpr (rocfft_version_major == 1 && rocfft_version_minor == 0 &&
```
This if-statement could be converted to `#if` directive. I preferred to use this style instead, but I am open to change it

s-Nick commented on src/dft/backends/rocfft/commit.cpp:
```diff
@@ -259,12 +260,37 @@ class rocfft_commit final : public dft::detail::commit_impl<prec, dom> {
+        // Link to rocFFT issue: https://github.com/ROCm/rocFFT/issues/507
+        if constexpr (rocfft_version_major == 1 && rocfft_version_minor == 0 &&
+                      (rocfft_version_patch > 22 && rocfft_version_patch < 31)) {
+            if (dom == dft::domain::COMPLEX && dimensions > 2) {
```
This change should allow `in_place` routine to run
```suggestion
            if (dom == dft::domain::COMPLEX &&  config_values.placement == dft::config_value::NOT_INPLACE && dimensions > 2) {
```

s-Nick committed:
Address PR comments

Removing rocFFT version check during configurations.
Update checks on dft command for rocFFT version with known issue.

Rbiessy commented on src/dft/backends/rocfft/CMakeLists.txt:
```diff
@@ -62,6 +62,7 @@ find_path(
   NO_DEFAULT_PATH
   REQUIRED
 )
+
```
Might as well remove this new line to reduce noise

s-Nick committed:
Remove extra line

lhuot commented on src/dft/backends/rocfft/descriptor.cpp:
```diff
@@ -22,9 +22,7 @@
-namespace oneapi {
-namespace mkl {
-namespace dft {
+namespace oneapi::mkl::dft::detail {
```
It seems the namespace in this file is not consistent in all backends.
With this change rocFFT, oneMKL CPU, oneMKL GPU have the oneapi::mkl::dft::detail namespace but cuFFT and portFFT have oneapi::mkl::dft namespace.
I know this is unrelated to the original issue, but since you changed one of them, could you please make them consistent across all backends?

s-Nick replied:
Thank you for your review @lhuot 
No problem, I updated them in [aed1f63](https://github.com/oneapi-src/oneMKL/pull/563/commits/aed1f63fea977b65a5c2da785b5405e4d1ae8850)

s-Nick committed:
Update namespace for cuFFT and portFFT descriptor

Update namespace to make it consistent across all backends.

raphael-egan commented on src/dft/backends/rocfft/commit.cpp:
```diff
@@ -259,12 +260,30 @@ class rocfft_commit final : public dft::detail::commit_impl<prec, dom> {
+        if constexpr (rocfft_version_major == 1 && rocfft_version_minor == 0 &&
+                      (rocfft_version_patch > 22 && rocfft_version_patch < 31)) {
+            if (dom == dft::domain::COMPLEX &&
+                config_values.placement == dft::config_value::NOT_INPLACE && dimensions > 2) {
```
Are we certain this is limited to `dimensions > 2`? Is a 2D case like
`lengths = {A, B}`
`inStrides = {B 1}`
`outStrides = {1, A}`
present in the test base?
Possibly relevant as well are 1D batch cases like
`length = N;`
`batch = M;`
`inStrides = 1; inDistance = N;`
`outStrides = M; outDistance = 1;`

s-Nick replied:
Thank you for your review @raphael-egan 
The 2D configuration above is not directly tested. If added, rocfft backend skips the tests with the following output:
```bash
Skipping test because: "oneMKL: DFT/commit: function is not implemented rocfft requires a stride to be divisible by all dimensions associated with smaller strides!"
```
If a configuration that respect inputs requirement is provided, tests are passed. Similar configurations are already in the test suite, so I don't think adding this test is a priority.

For 1D there isn't a test that reproduces what you suggested. I added it to check. It is computed correctly with the input above, although the `outDistance` is computed internally to perform the operation correctly. In any case, both configuration do not cause any issue with rocFFT.

Therefore, I wouldn't change the if-statement condition

raphael-egan replied:
Given that the exception is thrown by this project and not by rocFFT itself, I would recommend to add a comment like
`// rocFFT's functional status for problems like cfoA:B:1xB:1:A is unknown as of [commit hash] (due to project's implementation preventing testing thereof)`

Unless an explicit mention of that as a known limitation is found in rocFFT's documentation/release notes (I could not find any), it looks to me that the check triggering this exception to be thrown is present here to ensure consistency of subsequent stride-related implementation steps (sorting operations, in particular).

s-Nick replied:
I agree, added comment in [e84a5b2](https://github.com/oneapi-src/oneMKL/pull/563/commits/e84a5b28c44e6823e5a08d1a06e40b3171858666)

s-Nick committed:
Add comment to explain current implementation status
